File: 프로그래머스/lv2/12952. N－Queen/N－Queen.java
class Solution {
    int[] rows;
    public int solution(int n) {
        
        int answer = 0;
        rows = new int[n];
        answer = dfs(0,n);
        
        return answer;
    }
    
    public int dfs(int depth,int n){
        if (depth == n){
            return 1;
        }
        int cnt = 0;
        for (int i = 0; i < n; i++){
            rows[depth] = i;
            if (isValid(depth)){
                cnt += dfs(depth + 1, n);
            }
        }
        return cnt;
        
    }
    
    public boolean isValid(int x){
        
        for (int i = 0; i < x; i++){
            if(rows[x] == rows[i] || (Math.abs(x-i) == Math.abs(rows[x]-rows[i]))){
                return false;
            }
        }
        return true;
    }
    
}

File: archive/legacy/SWEA/D3/2806. N－Queen/N－Queen.py
def is_valid(x):
    for i in range(x):
        #이전 행과 현재 x행 놓은 퀸이 같은 열 또는 대각선에 있으면 안됌
        if (rows[x] == rows[i]) or abs(x-i) == abs(rows[x]- rows[i]):
            return False
    return True 

def dfs(depth):
    if depth == n:
        global answer
        answer += 1
        return 
    
    for i in range(n):
        rows[depth] = i
        if is_valid(depth):
            dfs(depth+1)


t = int(input())

for test_case in range(1,t+1):
    global answer 
    answer = 0
    n = int(input())
    rows = [-1] * (n)
    
    dfs(0)
    print(f'#{test_case} {answer}')